src/azul/plugins/__init__.py

```diff
@@ -744,6 +786,73 @@ class RepositoryFileDownload(metaclass=ABCMeta):
     #: ignore this.
     drs_uri: str | None
 
+    #: The file's size on disk, if known.
+    size: int | None = None
+
+    @classmethod
+    @abstractmethod
+    def from_hit(cls, hit: JSON) -> Self:
+        """
+        Instantiate this class from an entity aggregate document retrieved from
+        Elasticsearch.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def from_request(cls, uuid: str, params: Mapping[str, str]) -> Self:
+        """
+        Instantiate this class from an HTTP request with query parameters
+        populated from :meth:`to_request`.
+        """
+        fields = cls._fields_from_request(params)
+        return cls(uuid=uuid, **fields)
+
+    @classmethod
+    def _fields_from_request(cls, params: Mapping[str, str]) -> dict[str, Any]:
+        return {
+            'version': params['version'],
+            'name': params['fileName'],
+            'drs_uri': params.get('drsUri'),
+            'size': None
+        }
+
+    @abstractmethod
+    def to_request(self) -> dict[str, str]:
+        """
+        Serialize this class in the form of HTTP query parameters.
+        """
+        assert self.version is not None
+        return {
+            'version': self.version,
+            'fileName': self.name,
+            **({} if self.drs_uri is None else {'drs_uri': self.drs_uri})
+        }
+
+    @property
+    def mirror_object_key(self) -> str:
+        """
+        S3 object key for the copy of this file mirrored by Azul
+        """
+        digest, digest_type = self.digest()
+        assert all(c in string.hexdigits for c in digest), R(
+            'Expected a hexadecimal digest', digest)
+        return f'file/{digest.lower()}.{digest_type}'
```

Comment:
This also feels misplaced/brittle because it leaks a detail of the bucket layout which is a concern plugins should not be aware of.

In general, not every representation of an instance of a class needs to be produced/consumed by a method of said class. The factory class and builder patterns are good examples, and the DAO enterprise pattern is another one.